src/protocol/boolean/random_bits.rs

```diff
@@ -0,0 +1,170 @@
+    use crate::{
+        error::BoxError,
+        ff::{Field, Fp31, Fp32BitPrime},
+        protocol::{context::ProtocolContext, QueryId, RecordId},
+        secret_sharing::Replicated,
+        test_fixture::{
+            bits_to_field, make_contexts, make_world, validate_and_reconstruct, TestWorld,
+        },
+    };
+    use futures::future::try_join_all;
+    use rand::{distributions::Standard, prelude::Distribution};
+
+    const TEST_TRIES: usize = 100;
+
+    async fn random_bits<F: Field>(
+        ctx: [ProtocolContext<'_, Replicated<F>, F>; 3],
+        record_id: RecordId,
+    ) -> Result<(Vec<F>, F), BoxError>
+    where
+        Standard: Distribution<F>,
+    {
+        let [c0, c1, c2] = ctx;
+
+        // Execute
+        let result = try_join_all(vec![
+            RandomBits::execute(c0.bind(record_id), record_id),
+            RandomBits::execute(c1.bind(record_id), record_id),
+            RandomBits::execute(c2.bind(record_id), record_id),
+        ])
+        .await
+        .unwrap();
+
+        // just renaming the results for the ease of reading
+        let (r0_b, r0_p) = (result[0].0.clone(), &result[0].1);
+        let (r1_b, r1_p) = (result[1].0.clone(), &result[1].1);
+        let (r2_b, r2_p) = (result[2].0.clone(), &result[2].1);
+
+        // [r]_B must be the same bit lengths
+        assert_eq!(r0_b.len(), r1_b.len());
+        assert_eq!(r1_b.len(), r2_b.len());
+
+        // Reconstruct r_B from ([r_1]_p,...,[r_l]_p) bitwise sharings in F_p
+        let r_b = (0..r0_b.len())
+            .map(|i| {
+                let bit = validate_and_reconstruct(&r0_b[i], &r1_b[i], &r2_b[i]);
+                assert!(bit == F::ZERO || bit == F::ONE);
+                bit
+            })
+            .collect::<Vec<_>>();
+        let r_p = validate_and_reconstruct(r0_p, r1_p, r2_p);
+
+        Ok((r_b, r_p))
+    }
+
+    #[tokio::test]
+    pub async fn fp31() -> Result<(), BoxError> {
+        let world: TestWorld = make_world(QueryId);
+        let ctx = make_contexts::<Fp31>(&world);
+        let [c0, c1, c2] = ctx;
+
+        for i in 0..TEST_TRIES {
+            let record_id = RecordId::from(i);
+            let (r_b, r_p): (Vec<Fp31>, Fp31) =
+                random_bits([c0.clone(), c1.clone(), c2.clone()], record_id).await?;
+            // Base10 of r_b must equal r_p
+            assert_eq!(r_p, bits_to_field(&r_b));
+        }
+        Ok(())
+    }
+
+    #[tokio::test]
+    pub async fn fp_32bit_prime() -> Result<(), BoxError> {
+        let world: TestWorld = make_world(QueryId);
+        let ctx = make_contexts::<Fp32BitPrime>(&world);
+        let [c0, c1, c2] = ctx;
+
+        for i in 0..TEST_TRIES {
+            let record_id = RecordId::from(i);
+            let (r_b, r_p): (Vec<Fp32BitPrime>, Fp32BitPrime) =
+                random_bits([c0.clone(), c1.clone(), c2.clone()], record_id).await?;
+            // Base10 of r_b must equal r_p
+            assert_eq!(r_p, bits_to_field(&r_b));
```

Comment:
There are just 5 numbers that are more than this prime. Let's make sure we have some tests for them.